カレンダープログラムを作成

konzo-goriki opened the pull request:
カレンダープログラムを作成した。
課題として出されている
- 必要条件
- 歓迎要件
    - 今日の日付の部分の色が反転する

まで実装した。

konzo-goriki committed:
cal.rbを追加

カレンダープログラムcal.rbを追加した。

konzo-goriki committed:
今日の日付が反転するように修正

カレンダーを表示するとき、今日の日付が目立つように反転して表示させるため、cal.rbのdateメソッドを修正した。

maedana commented:
動作例にあるように

```
$ ./cal.rb -m 11
```
と、rubyをつけずにコマンドとして実行出来るようにしましょうー

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+end
+
+options = option
+month = to_i(options["-m"])
```
メソッドにしなくてもぼっち演算子で`options["-m"]&.to_i`ぐらいでいいかもですね

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+options = option
+month = to_i(options["-m"])
+year = to_i(options["-y"])
+years(month, year)
```
メソッド名は一部例外を除き動詞で始めるのがセオリーかなと思います。名詞のメソッドは単に値を返すメソッドかなって思っちゃいますね。でも実際は内部で出力処理してますし、そうとわかる命名がよさそうです。

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+
+def years(month=nil,year=nil)
+  month ||= Date.today.month
+  year ||= Date.today.year
```
メソッド定義部に初期値として入れちゃえばいいかなと思います。

konzo-goriki replied:
```
def years(month=Date.today.month,year=Date.today.year)
  #month ||= Date.today.month
  #year ||= Date.today.year
```

こういうことでしょうか？

このように初期値を入れてしまうと、現状の実装だとmonth,yearに与えられるのが
- 数値オブジェクト
- nilオブジェクト
のどちらかになるため、デフォルト値として指定したDate.today.monthおよびDate.today.yearが呼ばれず、その後Date.newでnilが渡されてしまったときにnot numericエラーとなってしまいます。

（デフォルト値が呼ばれることがないと言っているのにnilを与えているのはなんとなくです。すいません…。）

あるいは初期値として入れたうえで、メソッド呼び出し側を工夫するようになるでしょうか？

maedana replied:
確かにそうですねー。だったら初期値はメソッドに持たないようにして呼び出し側を
```ruby
month = options["-m"]&.to_i || Date.current.month
year = options["-y"]&.to_i || Date.current.year
```
にするとスッキリかなーと思いました

konzo-goriki replied:
ありがとうございます。

ご提案いただいたとおり、呼び出し側で処理するように修正しました。
なお、Date.currentはRailsの機能のようでしたのでDate.todayで処理しております。

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+  year ||= Date.today.year
+
+  puts("      #{month}月 #{year}")
+  ["日","月","火","水","木","金","土"].each {|w| print("#{w} ") }
```
これでいいですが、
`print ["日","月","火","水","木","金","土"].join(' ')`
という書き方もありますね。たぶんこっちのほうがRubyでよく見かける書き方かも?

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+  month ||= Date.today.month
+  year ||= Date.today.year
+
+  puts("      #{month}月 #{year}")
```
空白調整大変なので、
https://docs.ruby-lang.org/ja/latest/method/String/i/center.html
を使うという手もありますね

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+
+def date(month=nil,year=nil)
+  month ||= Date.today.month
+  year ||= Date.today.year
```
引数は同上です

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+  options
+end
+
+def years(month=nil,year=nil)
```
細かいこというと`=`の前後半角空白あけたいですね

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+  month ||= Date.today.month
+  year ||= Date.today.year
+
+  first_day = Date.new(year,month,1)
```
ここも細かいですが、カンマの前後半角空白あけたいです。

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+  first_day = Date.new(year,month,1)
+  last_day = Date.new(year,month,-1)
+
+  first_day.wday.times {print("　 ")}
```
これでいいですが、文字列の演算が出来るので、空白数は `'   ' * first_day.wday`みたいに算出することも出来ますね。ご参考までに

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+  last_day = Date.new(year,month,-1)
+
+  first_day.wday.times {print("　 ")}
+  (first_day..last_day).each do |day|
```
Date型をRangeで回すのいいですね :+1:

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+
+  first_day.wday.times {print("　 ")}
+  (first_day..last_day).each do |day|
+    unless day == Date.today then
```
then いらないですね。あと否定条件って基本は読みにくいと思ってて、特に今回のように`else`もあるなら肯定条件から書くようにしましょう

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+  first_day.wday.times {print("　 ")}
+  (first_day..last_day).each do |day|
+    unless day == Date.today then
+      print("#{day.strftime("%e ")}")
```
`"#{ }"`するまでもなく、`day.strftime("%e ")`の時点で文字列が返ってるのでは

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+    unless day == Date.today then
+      print("#{day.strftime("%e ")}")
+    else
+      print("#{day.strftime("\e[7m%e\e[0m ")}")
```
同上です

maedana commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,53 @@
+    else
+      print("#{day.strftime("\e[7m%e\e[0m ")}")
+    end
+    print("\n") if day.saturday?
```
`saturday?`使って判定しているのリーダブルでよいです :+1:

konzo-goriki committed:
shebangを追加

cal.rbにshebangを追加した。

konzo-goriki committed:
メソッド名を修正

cal.rbのメソッド"years"を"print_top_of_calendar","date"を"print_body_of_calendar"とメソッド名を修正した。

konzo-goriki committed:
to_iメソッドを消去

引数を数値オブジェクトに変換するメソッドto_iを消去し、ぼっち演算子を利用した実装に修正した。

konzo-goriki committed:
カレンダー上部の表示方法を修正

cal.rbのprint_top_of_calendarメソッドで年月、曜日の表示方法を修正した。

konzo-goriki committed:
カレンダー日付部分の表示方法を修正

cal.rbのprint_body_of_calendarメソッドで
- 不要な式展開を取り除く
- 空白を文字列の演算で表示
- unless文による条件分岐をif文による条件分岐
となるように修正した。

konzo-goriki commented:
@maedana 

確認とコメントありがとうございました！

基本的にご指摘いただいたとおりに修正しました。
一箇所、メソッド定義で初期値を与える実装についてコメント、質問させていただいております。

再度確認をよろしくお願いします。

konzo-goriki committed:
コマンドオプションをメソッド呼び出し側で処理

cal.rbでカレンダー表示用メソッドで使用するオプション引数の読み込み処理を
メソッド呼び出し側で処理するように変更し、不要なデフォルト引数を消去した。

konzo-goriki commented:
@maedana 

ご指摘いただいた箇所の修正をしました。
再度確認をよろしくお願いします。
